File: cp-book/ch2/lineards/bitmanip/_11173_GreyCodes.py
# Problem name: 11173 Grey Codes
# Problem url: https://uva.onlinejudge.org/external/111/11173.pdf
# Author: Andrey Yemelyanov

import sys
import math

def readline():
	return sys.stdin.readline().strip()

def main():
	n_tests = int(readline())
	for i in range(n_tests):
		n, k = [int(x) for x in readline().split()]
		print(to_decimal(generate_grey_code(n, k)))

def to_decimal(code):
	decimal = 0
	for i in range(len(code)):
		if code[i] == 1:
			decimal += pow(2, len(code) - 1 - i)
	return decimal

def generate_grey_code(n, k):
	code = []
	generate_grey_code_rec(n, k, code)
	return code

def generate_grey_code_rec(n, k, code):
	if n == 1:
		code.append(k)
		return
	if k < pow(2, n) / 2: # upper half
		code.append(0)
		generate_grey_code_rec(n - 1, k, code)
	else:
		code.append(1)
		generate_grey_code_rec(n - 1, pow(2, n) - 1 - k, code)

if __name__=="__main__":
    main()

File: cp-book/ch2/lineards/bitmanip/_11173_GreyCodes.java
import java.util.*;
import static java.lang.Math.*;
import java.util.stream.*;

/*
Problem name: 11173 Grey Codes
Problem url: https://uva.onlinejudge.org/external/111/11173.pdf
Author: Andrey Yemelyanov
*/
public class _11173_GreyCodes {
  public static void main(String[] args){
    Scanner s = new Scanner(System.in);
    int nTests = s.nextInt();
    while(nTests-- > 0){
      int n = s.nextInt(); int k = s.nextInt();
      System.out.println(toDecimal(generateGreyCode(n, k)));
    }
  }

  static long toDecimal(List<Integer> code){
    long decimal = 0;
    for(int i = 0; i < code.size(); i++){
      if(code.get(i) == 1){
        decimal += pow(2, code.size() - 1 - i);
      }
    }
    return decimal;
  }

  static List<Integer> generateGreyCode(int n, int k){
    List<Integer> code = new ArrayList<>();
    generateGreyCode(n, k, code);
    return code;
  }

  static void generateGreyCode(int n, int k, List<Integer> code){
    if(n == 1){
      code.add(k);
      return;
    }
    if(k < (int)pow(2, n) / 2){
      code.add(0);
      generateGreyCode(n - 1, k, code);
    }else{
      code.add(1);
      generateGreyCode(n - 1, (int)pow(2, n) - 1 - k, code);
    }
  }
}
